iceoryx_dust/test/moduletests/test_vocabulary_span.cpp

```diff
@@ -0,0 +1,347 @@
+
+    ASSERT_EQ(vector.data(), static_sut.data());
+    ASSERT_EQ(vector.size(), static_sut.size());
+    for (size_t i = 0; i < static_sut.size(); ++i)
+    {
+        EXPECT_EQ(vector[i], static_sut[i]);
+    }
+}
+
+TEST(span_test, NewConstSpanFromConstIoxUninitializedArrayContainsSameData)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "86ee3995-4267-4650-b1c4-4db8f5cf154b");
+    constexpr uint64_t CAPACITY{6U};
+    iox::UninitializedArray<int, CAPACITY> uninitializedArray;
+    uninitializedArray[0] = 60;
+    uninitializedArray[1] = 41;
+    uninitializedArray[2] = 21;
+    uninitializedArray[3] = 32;
+    uninitializedArray[4] = 53;
+    uninitializedArray[5] = 74;
+
+    span<const int> const_sut(uninitializedArray);
+
+    ASSERT_EQ(uninitializedArray.begin(), const_sut.data());
+    ASSERT_EQ(uninitializedArray.capacity(), const_sut.size());
+    for (size_t i = 0; i < const_sut.size(); ++i)
+    {
+        EXPECT_EQ(uninitializedArray[i], const_sut[i]);
+    }
+}
+
+TEST(span_test, NewStaticSpanFromConstIoxUninitializedArrayContainsSameData)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "c6a3f7d2-dbab-4c9f-a405-6ee2cc3f4457");
+    constexpr uint64_t CAPACITY{6U};
+    iox::UninitializedArray<int, CAPACITY> uninitializedArray;
+    uninitializedArray[0] = 66;
+    uninitializedArray[1] = 44;
+    uninitializedArray[2] = 22;
+    uninitializedArray[3] = 33;
+    uninitializedArray[4] = 55;
+    uninitializedArray[5] = 77;
+
+    span<const int, 6> static_sut(uninitializedArray.begin(), uninitializedArray.capacity());
+
+    ASSERT_EQ(uninitializedArray.begin(), static_sut.data());
+    ASSERT_EQ(uninitializedArray.capacity(), static_sut.size());
+    for (size_t i = 0; i < static_sut.size(); ++i)
+    {
+        EXPECT_EQ(uninitializedArray[i], static_sut[i]);
+    }
+}
+
+TEST(span_test, CheckFrontOfSpanIfItReturnsTheElementAtIndex0)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "57b2f67f-79c1-4c1e-a305-f4665283c474");
+    static constexpr int arr[] = {1, 6, 1, 8, 0};
+    constexpr span<const int> span(arr);
+    static_assert(&arr[0] == &span.front(), "span.front() does not refer to the same element as arr[0]");
+}
+
+TEST(span_test, CheckConstexprIterOfSpan)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8764fcfb-27df-4f39-b8cd-56bf881db382");
+    static constexpr int arr[] = {1, 6, 1, 8, 0};
+    IOX_MAYBE_UNUSED constexpr span<const int> span(arr);
+
+    // Explicitly not use EXPECT_TRUE here to be able to execute the test case during compile-time
+    // 'static_assert' is not possible as not being supported with GCC5
+    assert(1 == span.begin()[0]);      // First element needs to be '1'
+    assert(1 == *(span.begin() += 0)); // First element needs to be '1'
+    assert(6 == *(span.begin() += 1)); // Second element needs to be '6'
+
+    assert(1 == *((span.begin() + 1) -= 1)); // First element needs to be '1'
+    assert(6 == *((span.begin() + 1) -= 0)); // Second element needs to be '6'
```

Comment:
In this case please use `static_assert` and disable the test for gcc5 with an ifdef

Reply:
@elBoberido IIRC for some reason it's not only GCC5 but also one of the sanitiser jobs wasn't able to build this `constexpr`. I assume that has something to do with the optimization levels. Would you be ok with keeping the current approach?

Reply:
@mossmaurice `assert` is only a runtime check and `assert(false)` will happily compile. Do you know which job had problems? This indicates that something is wrong with the constexpr.

Reply:
@elBoberido As I stated in several threads above, `assert`s can be executed both during compile- and run-time, see [this Godbolt snippet](https://godbolt.org/z/13dThfbd9). If you don't mind I would go ahead and merge it as it is.

Reply:
@mossmaurice the linked example clearly shows that it is runtime only. The code compiles and only during runtime the assertion is checked. If it would be compile time, the output would be similar to this
```
<source>: In function 'int main()':
<source>:12:19: error: static assertion failed
   12 |     static_assert(false);
      |                   ^~~~~
ASM generation compiler returned: 1
<source>: In function 'int main()':
<source>:12:19: error: static assertion failed
   12 |     static_assert(false);
      |                   ^~~~~
Execution build compiler returned: 1
```
but with the `assert` it is
```
ASM generation compiler returned: 0
Execution build compiler returned: 0
Program returned: 139
output.s: /app/example.cpp:9: int main(): Assertion `false' failed.
 compile-time
```
I don't know why the string `compile-time` time occurs in the output. It is also there when there are no assertions. The assert itself occurred at runtime which is indicated by `Program returned: 139`

Reply:
@mossmaurice I take (almost) everything back I said about `assert` and `only runtime`. With C++14 it can indeed be used in a constexpr and it still works when the `assert` can be evaluated to `true` at compile time. But in your example it has lead to a runtime check and the string `compile-time` comes from the `std::cout`. I overlooked that before.

This godbolt snippet shows how it works with constexpr.
https://godbolt.org/z/s83TT54d9